src/macaron/artifact/maven.py

```diff
@@ -0,0 +1,203 @@
+
+    @classmethod
+    def from_package_url(cls, package_url: PackageURL) -> Self | None:
+        """Create a Maven artifact from a PackageURL.
+
+        Parameters
+        ----------
+        package_url : PackageURL
+            The PackageURL identifying a Maven artifact.
+
+        Returns
+        -------
+        Self | None
+            A Maven artifact, or ``None`` if the PURL is not a valid Maven artifact PURL, or if
+            the artifact type is not supported.
+            For supported artifact types, see :class:`MavenArtifactType`.
+        """
+        if not package_url.namespace:
+            return None
+        if not package_url.version:
+            return None
+        if package_url.type != "maven":
+            return None
+        maven_artifact_type = None
+        for artifact_type in MavenArtifactType:
+            if artifact_type.purl_qualifiers == package_url.qualifiers:
+                maven_artifact_type = artifact_type
+                break
+        if not maven_artifact_type:
+            return None
+        return cls(
+            group_id=package_url.namespace,
+            artifact_id=package_url.name,
+            version=package_url.version,
+            artifact_type=maven_artifact_type,
+        )
+
+    @classmethod
+    def from_artifact_name(
+        cls,
+        artifact_name: str,
+        group_id: str,
+        version: str,
+    ) -> Self | None:
+        """Create a Maven artifact given an artifact name.
+
+        The artifact type is determined based on the naming pattern of the artifact.
+
+        Parameters
+        ----------
+        artifact_name : str
+            The artifact name.
+        group_id : str
+            The group id.
+        version : str
+            The version
+
+        Returns
+        -------
+        Self | None
+            A Maven artifact, or ``None`` if the PURL is not a valid Maven artifact PURL, or if
+            the artifact type is not supported.
+            For supported artifact types, see :class:`MavenArtifactType`.
+        """
+        for maven_artifact_type in MavenArtifactType:
+            pattern = maven_artifact_type.filename_pattern.format(
+                artifact_id="(.*)",
+                version=version,
+            )
+            match_result = re.search(pattern, artifact_name)
+            if not match_result:
+                continue
+            artifact_id = match_result.group(1)
+            return cls(
+                group_id=group_id,
+                artifact_id=artifact_id,
+                version=version,
+                artifact_type=maven_artifact_type,
+            )
+        return None
+
+
+class MavenSubjectPURLMatcher:
+    """A matcher matching a PURL identifying a Maven artifact to a provenance subject."""
+
+    @staticmethod
+    def get_subject_in_provenance_matching_purl(
+        provenance_payload: InTotoPayload, purl: PackageURL
+    ) -> InTotoV01Subject | InTotoV1ResourceDescriptor | None:
```

Comment:
I don't think this method returns `InTotoV1ResourceDescriptor` instances. 
```
artifact_subjects = extract_build_artifacts_from_witness_subjects(provenance_payload)
```
The returned subject is an element in `artifact_subjects` whose type is `list[InTotoV01Subject]` 
https://github.com/oracle/macaron/blob/12066e31b59e64ecad29c59c703515083baf7468/src/macaron/slsa_analyzer/provenance/witness/__init__.py#L124

Reply:
Hmm I think it's because we are following the protocol here https://github.com/oracle/macaron/blob/7fa98d6e8118dca219f927cfaf123a368c54fd54/src/macaron/slsa_analyzer/provenance/intoto/__init__.py#L128

If that's the case then I don't think there is any issue.

Reply:
Yeah, the function signature of `get_subject_in_provenance_matching_purl` returns a union of multiple different subject types since it is meant for extracting the subject of any provenance. This will probably require some rethinking later on and I am not 100% that this interface is ideal. As you have seen the only provenance type that we are supporting is Witness, which is still in-toto v0.1.